### PR TITLE
Add ArchivePortfolioCommandHandler

### LINKED ISSUE #20: Create ApplicationContext

### PR DESCRIPTION
**Pull request by shailscript**
Partially resolution for #10
Continues after #16

### CODE REVIEW

File: libs/portfolio/domain/src/lib/exception/portfolio-already-archived.exception.ts [hunk 1]
```diff
@@ -1,4 +1,6 @@
-export class PortfolioAlreadyArchivedException extends Error {
+import { AggregateException } from '@pl-oss/domain-core';
+
+export class PortfolioAlreadyArchivedException extends AggregateException {
```
Comment (venusvavadiya) [inline libs/portfolio/domain/src/lib/exception/portfolio-already-archived.exception.ts:3]: Please rename the file and directory as `aggregate-exception`.
```suggestion
export class PortfolioAlreadyArchivedAggregateException extends AggregateException {
```

File: libs/portfolio/domain/src/lib/exception/portfolio-not-archived.exception.ts [hunk 2]
```diff
@@ -1,4 +1,6 @@
-export class PortfolioNotArchivedException extends Error {
+import { AggregateException } from '@pl-oss/domain-core';
+
+export class PortfolioNotArchivedException extends AggregateException {
```
Comment (venusvavadiya) [inline libs/portfolio/domain/src/lib/exception/portfolio-not-archived.exception.ts:3]:
```suggestion
export class PortfolioNotArchivedAggregateException extends AggregateException {
```

File: libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    jest.spyOn(testPortfolioAggregateRepository, 'save').mockResolvedValue(undefined);
+    jest.spyOn(testPortfolioAggregate, 'archive');
+  });
+
+  afterEach(jest.clearAllMocks);
+
+  describe('handle', () => {
+    it('should archive portfolioAggregate if unarchived', async () => {
+      testPortfolioAggregate.id = id;
+      testPortfolioAggregate.isArchived = false;
```
Comment (venusvavadiya) [inline libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts:36]: Do not test what `archive` does. Testing if it is called or not is sufficient. The responsibility of `archive` is already tested in the aggregate's spec file.

```suggestion
```

File: libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      expect(testPortfolioAggregateRepository.getById).toHaveBeenNthCalledWith(1, id);
+
+      expect(testPortfolioAggregate.archive).toHaveBeenCalledTimes(1);
+      expect(testPortfolioAggregate.archive).toHaveBeenNthCalledWith(1);
+
+      expect(testPortfolioAggregateRepository.save).toHaveBeenCalledTimes(1);
+      expect(testPortfolioAggregateRepository.save)
+        .toHaveBeenNthCalledWith(1, testPortfolioAggregate);
+
+      expect(testPortfolioAggregate.isArchived).toBeTruthy();
```
Comment (venusvavadiya) [inline libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts:50]:
```suggestion
```

File: libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    });
+
+    it('should throw new PortfolioAlreadyArchivedException if archived', async () => {
+      testPortfolioAggregate.id = id;
+      testPortfolioAggregate.isArchived = true;
+
+      await expect(testArchivePortfolioCommandHandler.handle(testArchivePortfolioCommand))
+        .rejects
+        .toThrowError(new PortfolioAlreadyArchivedException());
+    });
```
Comment (venusvavadiya) [inline libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts:60]: Business logic is already tested for the Aggregate. It is assumed that we are letting Exceptions bubble.

```suggestion
```

**Review by venusvavadiya — CHANGES_REQUESTED** [review]

File: libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts (hunk 4 above)
Comment (shailscript) [reply libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts:50]: I was expecting this. Thanks!

File: libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts (hunk 5 above)
Comment (shailscript) [reply libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts]: Resolved in https://github.com/venusvavadiya/ticksbook-main/pull/18/commits/766759fe8436e29b646c76a3826a4da22113745a

File: libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts (hunk 4 above)
Comment (shailscript) [reply libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts]: Resolved in https://github.com/venusvavadiya/ticksbook-main/pull/18/commits/766759fe8436e29b646c76a3826a4da22113745a

File: libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts (hunk 3 above)
Comment (shailscript) [reply libs/portfolio/domain/src/lib/command-handlers/archive-portfolio.command-handler.spec.ts]: Resolved in https://github.com/venusvavadiya/ticksbook-main/pull/18/commits/766759fe8436e29b646c76a3826a4da22113745a

File: libs/portfolio/domain/src/lib/exception/portfolio-not-archived.exception.ts (hunk 2 above)
Comment (shailscript) [reply libs/portfolio/domain/src/lib/exception/portfolio-not-archived.exception.ts]: Resolved in https://github.com/venusvavadiya/ticksbook-main/pull/18/commits/edcba7bf84a94dfbfc5c31773c026132ceebe79c

File: libs/portfolio/domain/src/lib/exception/portfolio-already-archived.exception.ts (hunk 1 above)
Comment (shailscript) [reply libs/portfolio/domain/src/lib/exception/portfolio-already-archived.exception.ts]: Resolved in https://github.com/venusvavadiya/ticksbook-main/pull/18/commits/edcba7bf84a94dfbfc5c31773c026132ceebe79c

**Review by venusvavadiya — APPROVED** [review]

**Review by hashptl — APPROVED** [review]

**Review by aditipatelpro — APPROVED** [review]

**Review by ronakmistry36 — APPROVED** [review]

**Pull request merged**